sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/ier12.processing.service.ts

```diff
@@ -301,12 +318,235 @@ export class IER12ProcessingService {
+   */
+  private async eventCodeForCompletedApplicationWithAwardWithheldDueToRestriction(
+    currentDisbursementScheduleId: number,
+  ): Promise<ApplicationEventCode.DISW | ApplicationEventCode.DISS> {
+    // Check if any disbursement award (full amount or a partial)
+    // was withheld due to a restriction.
+    switch (
+      await this.disbursementValueService.hasAnyFullOrPartialAwardWithheldDueToRestriction(
+        currentDisbursementScheduleId,
+      )
+    ) {
+      case true:
+        return ApplicationEventCode.DISW;
+      case false:
+        return ApplicationEventCode.DISS;
+    }
+  }
+
+  /**
+   * Get application event code for an application with completed status
+   * with sent disbursement.
+   * @param currentDisbursementScheduleId current disbursement schedule id.
+   * @returns application event code.
+   */
+  private async eventCodeForCompletedApplicationWithSentDisbursement(
+    currentDisbursementScheduleId: number,
+  ): Promise<
+    | ApplicationEventCode.DISE
+    | ApplicationEventCode.DISW
+    | ApplicationEventCode.DISS
+  > {
+    // Check if the disbursement has any feedback error.
+    switch (
+      await this.disbursementScheduleErrorsService.hasFTDisbursementFeedbackErrors(
+        currentDisbursementScheduleId,
+      )
+    ) {
+      case true:
+        return ApplicationEventCode.DISE;
+      case false: {
+        return this.eventCodeForCompletedApplicationWithAwardWithheldDueToRestriction(
+          currentDisbursementScheduleId,
+        );
+      }
+    }
+  }
+
+  /**
+   * Get application event code for an application with completed status
+   * with pending disbursement and completed COE.
+   * @param studentId student id.
+   * @returns application event code.
+   */
+  private async eventCodeForCompletedApplicationWithPendingDisbursementAndCompletedCOE(
+    studentId: number,
+  ): Promise<ApplicationEventCode.DISR | ApplicationEventCode.COEA> {
+    switch (
+      await this.studentRestrictionService.hasActiveStopFullTimeDisbursement(
+        studentId,
+      )
+    ) {
+      case true:
+        return ApplicationEventCode.DISR;
+      case false:
+        return ApplicationEventCode.COEA;
+    }
+  }
+
+  /**
+   * Get application event code for an application with completed status
+   * with pending disbursement.
+   * @param currentDisbursementSchedule current disbursement schedule.
+   * @param studentId student id.
+   * @returns application event code.
+   */
+  private async eventCodeForCompletedApplicationWithPendingDisbursement(
+    currentDisbursementSchedule: DisbursementSchedule,
+    studentId: number,
+  ): Promise<
+    | ApplicationEventCode.COER
+    | ApplicationEventCode.COED
+    | ApplicationEventCode.DISR
+    | ApplicationEventCode.COEA
+  > {
+    if (currentDisbursementSchedule.coeStatus === COEStatus.completed) {
+      // Check if disbursement is not sent due to restriction.
+      if (
+        isSameOrAfterDate(
```

Comment:
I would say this logic and line 535 can be moved into `eventCodeForCompletedApplicationWithPendingDisbursementAndCompletedCOE`

Reply:
If I got your question correctly, here the intention is to check 
`eventCodeForCompletedApplicationWithPendingDisbursementAndCompletedCOE` only after
 ```
isSameOrAfterDate(
          currentDisbursementSchedule.disbursementDate,
          addDays(DISBURSEMENT_FILE_GENERATION_ANTICIPATION_DAYS),
        )
```
 is satisfied.